File: PERSISTENCE/Core Data/[Objective-C] Core Data/coredata_cars/coredata_cars/addCarViewController.m
//
//  carDetailViewController.m
//  coredata_cars
//
//  Created by Luis Andres Castillo Hernandez on 11/18/15.
//  Copyright © 2015 devbros. All rights reserved.
//

#import "addCarViewController.h"

#import <CoreData/CoreData.h>


@interface addCarViewController ()

@end

@implementation addCarViewController

@synthesize device;



//MARK:

-(NSManagedObjectContext *) managedObjectContext
{
    NSManagedObjectContext *context = nil;
    id delegate = [[UIApplication sharedApplication] delegate];
    
    if([delegate  performSelector:@selector(managedObjectContext)])
    {
        context = [delegate managedObjectContext];
    }
    
    return context;
    
}//eom



//MARK: View Loading
- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view.
    
    //the previous view wants to edit the information
    if (self.device)
    {
        //getting the car existing info
        NSString * currCarMake  = [self.device valueForKey:@"carMake"];
        NSString * currCarModel = [self.device valueForKey:@"carModel"];
        NSString * currCarYear  = [self.device valueForKey:@"carYear"];
        
        //updating the textfields with this info
        [self.carMake setText:currCarMake];
        [self.carModel setText:currCarModel];
        [self.carYear setText:currCarYear];
    }
}

//MARK: Memory
- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}


- (IBAction)saveCar:(UIButton *)sender
{
    NSManagedObjectContext * context = [self managedObjectContext];
    
    //saving existing entry
    if (self.device)
    {
        [self.device setValue:self.carMake.text forKey:@"carMake"];
        [self.device setValue:self.carModel.text forKey:@"carModel"];
        [self.device setValue:self.carYear.text forKey:@"carYear"];
    }
    //saving a new entry
    else
    {
        NSManagedObject *newDevice = [NSEntityDescription insertNewObjectForEntityForName:@"Device" inManagedObjectContext:context];
        
        [newDevice setValue:self.carMake.text forKey:@"carMake"];
        [newDevice setValue:self.carModel.text forKey:@"carModel"];
        [newDevice setValue:self.carYear.text forKey:@"carYear"];
    }

    //checking if error saving values occurred
    NSError * error = nil;
    if( ![context save:&error])
    {
        NSLog(@"%@ %@", error, [error localizedDescription]);
        
        [self showAlert:[error localizedDescription]];
    }
    else
    {
        // going back to previous view
        [self.navigationController popViewControllerAnimated:YES];
    }
    
}//eo-action

- (IBAction)dismissKeyboard:(id)sender
{
    
    
}//eo-action

//MARK: Alert
-(void)showAlert:(NSString *) message
{
    UIAlertController * alertCont = [UIAlertController alertControllerWithTitle:@"" message:message preferredStyle: UIAlertControllerStyleActionSheet];
    
    UIAlertAction * okayAction = [UIAlertAction actionWithTitle:@"Okay" style:(UIAlertActionStyle) UIAlertActionStyleDefault handler:nil];
    
    [alertCont addAction:okayAction];
    
    [self presentViewController:alertCont animated:true completion:nil];
}//eom


@end

File: PERSISTENCE/Core Data/[Swift] Core Data/coredata_cars/coredata_cars/addCarViewController.swift
//
//  ViewController.swift
//  coredata_cars
//
//  Created by Luis Andres Castillo Hernandez on 11/20/15.
//  Copyright © 2015 devbros. All rights reserved.
//

import UIKit
import CoreData

class addCarViewController: UIViewController, UITextFieldDelegate {

    @IBOutlet weak var carMakeTextField: UITextField!
    @IBOutlet weak var carModelTextField: UITextField!
    @IBOutlet weak var carYearTextField: UITextField!
    
    
    var carmake: String     = ""
    var carmodel: String    = ""
    var caryear: String     = ""
    
    var existingItem: NSManagedObject!
    
    
    override func viewDidLoad() {
        super.viewDidLoad()
        // Do any additional setup after loading the view, typically from a nib.
        
        
        //updating existing entry
        if(existingItem != nil)
        {
            carMakeTextField.text   = carmake
            carModelTextField.text  = carmodel
            carYearTextField.text   = caryear
        }
        
    }//ep

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }

    //MARK: Actions
    
    @IBAction func saveCarInfo(_ sender: AnyObject)
    {
        self.savedData()
    }//eoa

    
    //MARK: save
    func savedData()
    {
        let AppDel:AppDelegate = UIApplication.shared.delegate as! AppDelegate
        
        let theContext: NSManagedObjectContext = AppDel.managedObjectContext
        
        let theEnt = NSEntityDescription.entity(forEntityName: "CarList", in: theContext)
                
        //updating existing entry
        if(existingItem != nil)
        {
            let currCarMake     = carMakeTextField.text
            let currCarModel    = carModelTextField.text
            let currCarYear     = carYearTextField.text
            
            existingItem.setValue(currCarMake, forKey: "carMake")
            existingItem.setValue(currCarModel, forKey: "carModel")
            existingItem.setValue(currCarYear, forKey: "carYear")
        }
            //saving new entry
        else
        {
            let newitem = carModel(entity: theEnt!, insertInto: theContext)
            
            newitem.carMake     = carMakeTextField.text!
            newitem.carModel    = carModelTextField.text!
            newitem.carYear     = carYearTextField.text!
        }
        
        //going back to main view
        self.navigationController?.popViewController(animated: true)

    }//eom
    
    //MARK: textfield delegates
    func textFieldDidEndEditing(_ textField: UITextField)
    {
        textField.resignFirstResponder()
    }//eom
    
    func textFieldShouldReturn(_ textField: UITextField) -> Bool
    {
        textField.resignFirstResponder()
        
        if(textField == carMakeTextField)
        {
            carModelTextField.becomeFirstResponder()
        }
        else if (textField == carModelTextField)
        {
            carYearTextField.becomeFirstResponder()
        }
        else if (textField == carYearTextField)
        {
            self.savedData()
        }
        
        
        return true
    }//eom
    
}//eoc

